lib/ddtrace/configuration.rb

```diff
@@ -14,12 +14,26 @@ module Configuration
     # a threaded webserver), and we don't want their initialization to clash (for instance, starting two profilers...).
     #
     # Note that a Mutex **IS NOT** reentrant: the same thread cannot grab the same Mutex more than once.
-    # This means below we are careful not to nest calls to methods that grab the lock.
+    # This means below we are careful not to nest calls to methods that would trigger initialization and grab the lock.
     #
-    # Every method that directly or indirectly accesses/mutates @components should be holding the lock (through
+    # Every method that directly or indirectly mutates @components should be holding the lock (through
     # #safely_synchronize) while doing so.
-    COMPONENTS_LOCK = Mutex.new
-    private_constant :COMPONENTS_LOCK
+    COMPONENTS_WRITE_LOCK = Mutex.new
+    private_constant :COMPONENTS_WRITE_LOCK
+
+    # We use a separate lock when reading the @components, so that they continue to be accessible during initialization/
+    # reconfiguration. This was needed because we ran into several issues where we still needed to read the old
+    # components while the COMPONENTS_WRITE_LOCK was being held (see https://github.com/DataDog/dd-trace-rb/pull/1387
+    # and https://github.com/DataDog/dd-trace-rb/pull/1373#issuecomment-799593022 ).
+    #
+    # Technically on MRI we could get away without this lock, but on non-MRI Rubies, we may run into issues because
+    # we fall into the "UnsafeDCLFactory" case of https://shipilev.net/blog/2014/safe-public-construction/ .
+    # Specifically, on JRuby reads from the @components do have volatile semantics, and on TruffleRuby they do
```

Comment:
Crap, just noticed this should have been "do NOT have volatile semantics". I will correct this when I backport to the main branch.